searchlib/src/vespa/searchlib/features/internal_max_reduce_prod_join_feature.cpp

```diff
@@ -170,43 +154,120 @@ selectTypedExecutor(const IAttributeVector *attribute, IntegerVector vector, ves
+      _queryVectorKey()
+{
+}
+
+InternalMaxReduceProdJoinBlueprint::~InternalMaxReduceProdJoinBlueprint() = default;
+
+void
+InternalMaxReduceProdJoinBlueprint::visitDumpFeatures(const IIndexEnvironment &, IDumpFeatureVisitor &) const {
+}
+
+Blueprint::UP
+InternalMaxReduceProdJoinBlueprint::createInstance() const {
+    return std::make_unique<InternalMaxReduceProdJoinBlueprint>();
+}
+
+ParameterDescriptions
+InternalMaxReduceProdJoinBlueprint::getDescriptions() const {
+    return ParameterDescriptions().desc().attribute(ParameterDataTypeSet::int32OrInt64TypeSet(),
+                                                    ParameterCollection::ARRAY).string();
+}
+
+bool
+InternalMaxReduceProdJoinBlueprint::setup(const IIndexEnvironment &env, const ParameterList &params) {
+    _attribute = params[0].getValue();
+    _attrKey = createAttributeKey(_attribute);
+    _queryVector = params[1].getValue();
+    _queryVectorKey = make_queryvector_key(getBaseName(), _queryVector);
+    describeOutput("scalar", "Internal executor for optimized execution of reduce(join(A,Q,f(x,y)(x*y)),max)");
+    env.hintAttributeAccess(_attribute);
+    return true;
+}
+
+void
+InternalMaxReduceProdJoinBlueprint::prepareSharedState(const fef::IQueryEnvironment & env, fef::IObjectStore & store) const
+{
+    const IAttributeVector * attribute = lookupAndStoreAttribute(_attrKey, _attribute, env, store);
+    if (attribute == nullptr) return;
+
+    const fef::Anything * queryVector = env.getObjectStore().get(_queryVectorKey);
+    if (queryVector == nullptr) {
+        std::unique_ptr<IntegerVector> vector = createQueryVector(env.getProperties().lookup(_queryVector));
+        if (vector) {
+            store.add(_queryVectorKey, std::move(vector));
+        }
+    }
+}
 
 FeatureExecutor &
 InternalMaxReduceProdJoinBlueprint::createExecutor(const IQueryEnvironment &env, vespalib::Stash &stash) const
 {
-    const IAttributeVector *attribute = env.getAttributeContext().getAttribute(_attribute);
+    const IAttributeVector * attribute = lookupAttribute(_attrKey, _attribute, env);
     if (attribute == nullptr) {
         LOG(warning, "The attribute vector '%s' was not found in the attribute manager, "
-                "returning executor with default value.",
-            _attribute.c_str());
+                "returning executor with default value.", _attribute.c_str());
         return stash.create<SingleZeroValueExecutor>();
     }
-    Property prop = env.getProperties().lookup(_query);
-    if (prop.found() && !prop.get().empty()) {
-        IntegerVector vector;
-        WeightedSetParser::parse(prop.get(), vector);
-        if (!vector.getVector().empty()) {
+    const fef::Anything * queryVectorArg = env.getObjectStore().get(_queryVectorKey);
+    if (queryVectorArg != nullptr) {
+        // Vector is not copied as it is safe in ObjectStore
+        return selectExecutor<const IntegerVector &>(attribute, *dynamic_cast<const IntegerVector *>(queryVectorArg), stash);
```

Comment:
when using const references and unique pointers as alternatives, you should probably use universal references and std::forward instead of 'by value' and std::move...

Reply:
Done.